### PR TITLE
DAOS-16111 rebuild: uniform identifier in logs part 2 (migrate)

### PR DESCRIPTION
**Pull request by kccain**
When rebuild emits log messages, include a uniform rebuild operation
identifier. This change adjusts existing logging for rebuild migrate
activities. A previous patch added the same operation identifier
in log messages by the PS leader and storage engine scan activities.

Reminder, the baseline format (defined in DF_RB) for the uniform
identifier is:
"rb=" DF_UUID "/%u/%u/%s"
that corresponds to:
<pool_uuid>/<rebuild_ver>/<rebuild_gen>/<opcode_string>

This change adds DP_RB_OMI, DP_RB_MPT, and DP_RB_MRO macros that accept
struct obj_migrate_in *omi, struct migrate_pool_tls *mpt, and
struct migrate_one *mro, respectively, to provide the values needed for
the DF_RB logging format. And the patch applies them throughout the
existing logging performed in migrate activities.

Also in this change is some logic added to rebuild_leader_status_check():
a new function, warn_for_slow_engine_updates(). This allows a PS leader
engine to emit warnings when an engine is not reporting its rebuild
progress (via IV) for a long amount of time, making it easier for an
engineer to identify what engine(s) may be causing a stuck rebuild.
The warning messages are throttled to avoid too many log file entries.

Features: rebuild
Allow-unstable-test: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by kccain** [comment]
@liuxuezhao I have this follow-on to PR https://github.com/daos-stack/daos/pull/14383 (which is awaiting landing to master branch).

While I try to get this PR through CI (including with Features: rebuild), if you have some time can you start to review this change? Thanks.

**Comment by kccain** [comment]
Testing has gone well in Jenkins build 12 (Features: rebuild) and build 14 (pr test_nvme_pool_extend).

Ready for review.

The patch has a conflict with latest master (src/object/srv_obj_migrate.c), that I will address at the same time as resolving reviewer comments.

**Comment by kccain** [comment]
@liuxuezhao and @wangshilong will you have a chance to review this code change? It is a follow-on to https://github.com/daos-stack/daos/pull/14383 that I would like to both be included in an upcoming release. Thanks!

**Comment by kccain** [comment]
> @liuxuezhao and @wangshilong will you have a chance to review this code change? It is a follow-on to #14383 that I would like to both be included in an upcoming release. Thanks!

I really think that patches like this can help developers when we need to debug difficult problems, and those at scale, by making it easier to search for logging related to a specific rebuild operation. 

I guess the next opportunity to get the part 1 (already landed) patch https://github.com/daos-stack/daos/pull/14383 , and this one, would be DAOS 2.8 community release? Unless we're possibly too late even for that?

File: src/rebuild/srv.c [hunk 1]
```diff
@@ -598,6 +614,38 @@ enum {
...
+		double   tu = now - rgt->rgt_servers_last_update[i];
+		d_rank_t r  = rgt->rgt_servers[i].rank;
+
+		if (rgt->rgt_servers[i].scan_done && rgt->rgt_servers[i].pull_done)
+			continue;
+
+		/* Warn if it's been a while (> 30 seconds) since receiving an update from the rank.
+		 * And throttle warnings - do not print more than once per 2 minutes period.
+		 */
+		if ((tu > 30) && (tw > 120)) {
```
Comment (liuxuezhao) [inline src/rebuild/srv.c:636]: if (tw < 120) then need not traverse the array and compare, can return ahead.

File: src/rebuild/srv.c [hunk 2]
```diff
@@ -189,6 +189,20 @@ rebuild_leader_set_status(struct rebuild_global_pool_tracker *rgt,
 		status->pull_done = 1;
 }
 
+static void
+rebuild_leader_set_update_time(struct rebuild_global_pool_tracker *rgt, d_rank_t rank)
+{
+	int i;
+
+	for (i = 0; i < rgt->rgt_servers_number; i++) {
+		if (rgt->rgt_servers[i].rank == rank) {
```
Comment (liuxuezhao) [inline src/rebuild/srv.c:198]: can consider to sort rgt_servers[i].rank and do binary search the rank here

**Review by liuxuezhao — COMMENTED** [review]
please rebase, thx

**Comment by kccain** [comment]
> Test stage Functional Hardware Large completed with status FAILURE. https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-14547/17/execution/node/1432/log

erasurecode/multiple_target_failure.py test failure is an instance of known issue https://daosio.atlassian.net/browse/DAOS-16766

**Comment by kccain** [comment]
ping reviewers

File: src/rebuild/srv.c [hunk 3]
```diff
@@ -637,6 +701,39 @@ enum {
...
+	double tw     = now - rgt->rgt_last_warn; /* time since last warning logged */
+	bool   warned = false;
+
+	/* Throttle warnings to not more often than once per 2 minutes */
+	if (tw < 120)
+		return;
+
+	/* Warn for ranks not done and that haven't provided updates in a while (> 30 sec) */
+	for (i = 0; i < rgt->rgt_servers_number; i++) {
+		double   tu = now - rgt->rgt_servers_last_update[i];
```
Comment (wangshilong) [inline src/rebuild/srv.c:718]: I am not sure but it looks 30 sec is too aggressive...

Comment (liuxuezhao) [reply src/rebuild/srv.c:718]: target server reports progress every 2 second, so 30 sec here looks not too aggressive

**Review by liuxuezhao — APPROVED** [review]

**Review by wangshilong — APPROVED** [review]

**Pull request merged**